### PR TITLE
[mako] Add Mako integration

### PR DESCRIPTION
**Pull request by wklken**
- Maybe mako template is not so popular as jinja2
- We use it as default template of our framework , so here the contrib

### CODE REVIEW

File: ddtrace/contrib/mako/patch.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+from .constants import DEFAULT_TEMPLATE_NAME
+
+
+def patch():
+    if getattr(mako, '__datadog_patch', False):
+        # already patched
+        return
+    setattr(mako, '__datadog_patch', True)
+
+    Pin(service="mako", app="mako", app_type=http.TEMPLATE).onto(Template)
```
Comment (brettlangdon) [inline ddtrace/contrib/mako/patch.py]:
```suggestion
    Pin(service='mako', app='mako', app_type=http.TEMPLATE).onto(Template)
```

File: ddtrace/contrib/mako/patch.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    _u(mako.template.Template, 'render_context')
+
+
+def _wrap_render(wrapped, instance, args, kwargs):
+    pin = Pin.get_from(instance)
+    if not pin or not pin.enabled():
+        return wrapped(*args, **kwargs)
+
+    template_name = instance.filename or DEFAULT_TEMPLATE_NAME
+    print(func_name(wrapped))
```
Comment (brettlangdon) [inline ddtrace/contrib/mako/patch.py]:
```suggestion
```

File: ddtrace/monkey.py [hunk 3]
```diff
@@ -46,6 +46,7 @@
     'vertica': True,
     'molten': True,
     'jinja2': True,
+    'mako': False,
```
Comment (brettlangdon) [inline ddtrace/monkey.py]:
```suggestion
    'mako': True,
```

Comment (brettlangdon) [reply ddtrace/monkey.py]: We can enable this by default, that is what we'd like to do for all integrations :)

File: tests/contrib/mako/test_mako.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+from mako.lookup import TemplateLookup
+from mako.runtime import Context
+
+from ddtrace import Pin
+from ddtrace.contrib.mako import patch, unpatch
+from ddtrace.compat import StringIO, to_unicode
+from tests.test_tracer import get_dummy_tracer
+
+TEST_DIR = os.path.dirname(os.path.realpath(__file__))
+TMPL_DIR = os.path.join(TEST_DIR, "templates")
```
Comment (brettlangdon) [inline tests/contrib/mako/test_mako.py]:
```suggestion
TMPL_DIR = os.path.join(TEST_DIR, 'templates')
```

File: tests/contrib/mako/test_mako.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        patch()
+        self.tracer = get_dummy_tracer()
+        Pin.override(Template, tracer=self.tracer)
+
+    def tearDown(self):
+        unpatch()
+
+    def test_render(self):
+        # render
+        t = Template("Hello ${name}!")
```
Comment (brettlangdon) [inline tests/contrib/mako/test_mako.py]: `s/"/'/g`

Please be sure to use single quotes for this file.

File: tests/contrib/mako/test_mako.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        eq_(t.render(name="mako"), "Hello mako!")
+
+        spans = self.tracer.writer.pop()
+        eq_(len(spans), 1)
```
Comment (brettlangdon) [inline tests/contrib/mako/test_mako.py]: Please use `unittest.TestCase`s assertion functions, `self.assertEqual()`

sorry, I know looking at our other tests can be confusing, not everything is 100% consistent right now.

File: tests/contrib/mako/test_mako.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+        for span in spans:
```
Comment (brettlangdon) [inline tests/contrib/mako/test_mako.py]: We should only have 1 span here, can we update to be `self.assertEqual(spans[0].service, 'mako')`, etc?

File: tests/contrib/mako/test_mako.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+            eq_(span.get_tag("mako.template_name"), "<memory>")
+
+        eq_(spans[0].name, "mako.template.render")
+
+        # render_unicode
+        t = Template("Hello ${name}!")
+        eq_(t.render_unicode(name="mako"), to_unicode("Hello mako!"))
+        spans = self.tracer.writer.pop()
+        eq_(len(spans), 1)
+        eq_(spans[0].name, "mako.template.render_unicode")
```
Comment (brettlangdon) [inline tests/contrib/mako/test_mako.py]: we should also assert on `service`, `span_type`, etc here as well.

File: tests/contrib/mako/test_mako.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+        # render_context
+        t = Template('Hello ${name}!')
+        buf = StringIO()
+        c = Context(buf, name='mako')
+        t.render_context(c)
+        eq_(buf.getvalue(), "Hello mako!")
+        spans = self.tracer.writer.pop()
+        eq_(len(spans), 1)
+        eq_(spans[0].name, "mako.template.render_context")
```
Comment (brettlangdon) [inline tests/contrib/mako/test_mako.py]: Same as above, we should assert on `service`, `span_type`, etc here

File: tests/contrib/mako/test_mako.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+    def test_file_template(self):
+        tmpl_lookup = TemplateLookup(directories=[TMPL_DIR])
+        t = tmpl_lookup.get_template("template.html")
```
Comment (brettlangdon) [inline tests/contrib/mako/test_mako.py]: Should we be wrapping/tracing `get_template()`?

File: tests/contrib/mako/test_mako.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+        eq_(t.render(name="mako"), "Hello mako!\n")
+
+        spans = self.tracer.writer.pop()
+        eq_(len(spans), 1)
+
+        for span in spans:
```
Comment (brettlangdon) [inline tests/contrib/mako/test_mako.py]: Same here, we only have 1 span, the loop is not needed.

File: tests/contrib/mako/test_mako.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+            eq_(span.span_type, "template")
+            eq_(span.service, "mako")
+            eq_(span.get_tag("mako.template_name"), os.path.join(TMPL_DIR, "template.html"))
```
Comment (brettlangdon) [inline tests/contrib/mako/test_mako.py]: We are missing assertions for `span[0].name`, `span[0].resource`, etc

File: tests/contrib/mako/test_mako.py (hunk 8 above)
Comment (brettlangdon) [reply tests/contrib/mako/test_mako.py]: We also want to assert on the `spans[0].resource` value

File: tests/contrib/mako/test_mako.py (hunk 9 above)
Comment (brettlangdon) [reply tests/contrib/mako/test_mako.py]: We also want to assert on the `spans[0].resource` value

File: tests/contrib/mako/test_mako.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+        spans = self.tracer.writer.pop()
+        eq_(len(spans), 1)
+
+        for span in spans:
+            eq_(span.service, "mako")
+            eq_(span.span_type, "template")
+            eq_(span.get_tag("mako.template_name"), "<memory>")
+
+        eq_(spans[0].name, "mako.template.render")
```
Comment (brettlangdon) [inline tests/contrib/mako/test_mako.py]: We also want to assert on the `spans[0].resource` value

File: tox.ini [hunk 14]
```diff
@@ -206,6 +207,7 @@ deps =
     jinja28: jinja2>=2.8,<2.9
     jinja29: jinja2>=2.9,<2.10
     jinja210: jinja2>=2.10,<2.11
+    mako100: mako>=1.0.4,<1.1.0
```
Comment (brettlangdon) [inline tox.ini]: Why `1.0.4` and not `1.0.0`?

File: tox.ini [hunk 15]
```diff
@@ -66,6 +66,7 @@ envlist =
     grpc_contrib-{py27,py34,py35,py36}-grpc
     httplib_contrib-{py27,py34,py35,py36}
     jinja2_contrib-{py27,py34,py35,py36}-jinja{27,28,29,210}
+    mako_contrib-{py27,py34,py35,py36}-mako{100}
```
Comment (brettlangdon) [inline tox.ini]: Does this patching also happen to work for `mako<1.0.0` by any chance?

File: ddtrace/contrib/mako/__init__.py [hunk 16]
```diff
@@ -0,0 +1,12 @@
+from ..util import require_modules
```
Comment (brettlangdon) [inline ddtrace/contrib/mako/__init__.py]: We need to add a docstring here explaining the integration and how to use it.

Example: https://github.com/DataDog/dd-trace-py/blob/a0bdd4822127ce50e3776ca703e1266b0c36ab6d/ddtrace/contrib/flask/__init__.py#L1-L89

**Comment by wklken** [comment]
Thanks for the review @brettlangdon 
I will fix these later tomorrow:)

File: tox.ini (hunk 14 above)
Comment (wklken) [reply tox.ini]: I will check the early version

File: tox.ini (hunk 15 above)
Comment (wklken) [reply tox.ini]: I will check the early version

File: tests/contrib/mako/test_mako.py (hunk 10 above)
Comment (wklken) [reply tests/contrib/mako/test_mako.py]: The render takes most of the time of the hole render.

File: .circleci/config.yml [hunk 17]
```diff
@@ -1010,6 +1010,9 @@ workflows:
       - jinja2:
           requires:
             - flake8
+      - mako:
```
Comment (brettlangdon) [inline .circleci/config.yml]: Sorry, we want these to be listed alphabetically, can you move to be after `kombu` but before `molten` ?

File: .circleci/config.yml [hunk 18]
```diff
@@ -1112,6 +1115,7 @@ workflows:
             - httplib
             - integration
             - jinja2
+            - mako
```
Comment (brettlangdon) [inline .circleci/config.yml]: Same here, mind moving to between `kombu` and `molten`?

File: tox.ini [hunk 19]
```diff
@@ -66,7 +66,7 @@ envlist =
     grpc_contrib-{py27,py34,py35,py36}-grpc
     httplib_contrib-{py27,py34,py35,py36}
     jinja2_contrib-{py27,py34,py35,py36}-jinja{27,28,29,210}
-    mako_contrib-{py27,py34,py35,py36}-mako{100}
+    mako_contrib-{py27,py34,py35,py36}-mako{010,100}
```
Comment (brettlangdon) [inline tox.ini]: nice! glad this worked!

File: .circleci/config.yml (hunk 17 above)
Comment (wklken) [reply .circleci/config.yml]: sorry, I did not notice that.

**Comment by brettlangdon** [comment]
thanks @wklken!!!

I have this scheduled for our next release `0.20.0`, hopefully 1-2 weeks.

**Pull request merged**